File: app/src/main/java/com/vmb/financialcalculator/NumberTextWatcher.java
package com.vmb.financialcalculator;

import java.text.DecimalFormat;
import java.text.ParseException;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.EditText;

public class NumberTextWatcher implements TextWatcher {

    private DecimalFormat df;
    private DecimalFormat dfnd;
    private boolean hasFractionalPart;

    private EditText editText;

    public NumberTextWatcher(EditText editText) {
        df = new DecimalFormat("#,###.##");
        df.setDecimalSeparatorAlwaysShown(true);
        dfnd = new DecimalFormat("#,###");
        this.editText = editText;
        hasFractionalPart = false;
    }

    @SuppressWarnings("unused")
    private static final String TAG = "NumberTextWatcher";

    @Override
    public void afterTextChanged(Editable s) {
        editText.removeTextChangedListener(this);

        try {
            int inilen, endlen;
            inilen = editText.getText().length();
            String v = s.toString().replace(String.valueOf(df.getDecimalFormatSymbols().getGroupingSeparator()), "");
            Number n = df.parse(v);
            int cp = editText.getSelectionStart();
            if (hasFractionalPart) {
                editText.setText(df.format(n));
            } else {
                editText.setText(dfnd.format(n));
            }
            endlen = editText.getText().length();
            int sel = (cp + (endlen - inilen));
            if (sel > 0 && sel <= editText.getText().length()) {
                editText.setSelection(sel);
            } else {
                editText.setSelection(editText.getText().length() - 1);
            }
        } catch (NumberFormatException | ParseException ignored) {
        }
        editText.addTextChangedListener(this);
    }

    @Override
    public void beforeTextChanged(CharSequence s, int start, int count, int after) {
    }

    @Override
    public void onTextChanged(CharSequence s, int start, int before, int count) {
        hasFractionalPart = s.toString().contains(String.valueOf(df.getDecimalFormatSymbols().getDecimalSeparator()));
    }

    public static String GetStringInView(View dataEditText){
        String data = ((EditText) dataEditText).getText().toString();
        if (data.equals("")){
            data = "0";
        }
        data = data.replaceAll("\\s+","");
        return data;
    }
}

File: app/src/main/java/com/vmb/financialcalculator/NumberTextWatcher.kt
package com.vmb.financialcalculator

import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.EditText
import java.text.DecimalFormat
import java.text.ParseException

class NumberTextWatcher(editText: EditText) : TextWatcher {
    private val df: DecimalFormat
    private val dfnd: DecimalFormat
    private var hasFractionalPart: Boolean
    private val editText: EditText
    override fun afterTextChanged(s: Editable) {
        editText.removeTextChangedListener(this)
        try {
            val inilen: Int
            val endlen: Int
            inilen = editText.text.length
            val v = s.toString().replace(df.decimalFormatSymbols.groupingSeparator.toString(), "")
            val n = df.parse(v)
            val cp = editText.selectionStart
            if (hasFractionalPart) {
                editText.setText(df.format(n))
            } else {
                editText.setText(dfnd.format(n))
            }
            endlen = editText.text.length
            val sel = cp + (endlen - inilen)
            if (sel > 0 && sel <= editText.text.length) {
                editText.setSelection(sel)
            } else {
                editText.setSelection(editText.text.length - 1)
            }
        } catch (ignored: NumberFormatException) {
        } catch (ignored: ParseException) {
        }
        editText.addTextChangedListener(this)
    }

    override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
    override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
        hasFractionalPart = s.toString().contains(df.decimalFormatSymbols.decimalSeparator.toString())
    }

    companion object {
        private const val TAG = "NumberTextWatcher"
        fun GetStringInView(dataEditText: View): String {
            var data = (dataEditText as EditText).text.toString()
            if (data == "") {
                data = "0"
            }
            data = data.replace("\\s+".toRegex(), "")
            return data
        }
    }

    init {
        df = DecimalFormat("#,###.##")
        df.isDecimalSeparatorAlwaysShown = true
        dfnd = DecimalFormat("#,###")
        this.editText = editText
        hasFractionalPart = false
    }
}